sql: avoid integer overflow in generate_series

madelynnblue opened the pull request:
Fixes #18021

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/18066)
<!-- Reviewable:end -->

knz commented on pkg/sql/logictest/testdata/logic_test/generators:
```diff
@@ -234,3 +234,10 @@ SELECT * from generate_series(1, (select * from generate_series(1, 0)))
+SELECT GENERATE_SERIES(9223372036854775807::int, -9223372036854775807::int, -9223372036854775807::int)
+----
+9223372036854775807
+0
```
I'm not sure that's correct. the right bound for `generate_series` is inclusive.
If we can't reach that value at all, then the call should produce an error instead.

mjibson replied:
Good catch. Updated to fix this.

madelynnblue committed:
sql: avoid integer overflow in generate_series

While here, use the same code path for all the places we care about
int addition overflow.

Fixes #18021

knz commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 3 of 4 files at r1, 2 of 2 files at r2.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/18066#-:-KstLAlh2zwrj2-UuUWV:bnfp4nl)*
<!-- Sent from Reviewable.io -->
